### DISCUSSION TITLE
Request for help on MNIST RGM classifier

### DISCUSSION THREAD

**Discussion by Arun-Niranjan** [infer-actively/pymdp#338] [General]
Hello everyone,

I am attempting to implement a renormalised generative model MNIST image classifier as per the [scale free active inference paper](https://www.frontiersin.org/journals/network-physiology/articles/10.3389/fnetp.2025.1521963/full). I have a question on the paper I could use some help with if anyone has further context:

Question 1. Parameter Learning (Active)
Parameter learning updates occur according to `P(u)` defined in equation (6)
<img width="466" height="126" alt="equation_6" src="https://github.com/user-attachments/assets/03bc0ef6-077f-45dc-9b63-00aa1904ca93" />
where σ is a softmax (which reduces to a sigmoid in the case of a single number) and α is a precision hyperparameter controlling the sensitivity of parameter updates (a larger α means we are doing Bayesian Model Selection, a lower α corresponds to Bayesian Model Averaging)

My issue is with the free energy calculation in the case of a classifier. If I understand correctly, the cost(or utility) term is 0 (a classifier has no preferences on what it sees, we are in pure perception mode). therefore the free energy is equivalent to the mutual information. the mutual information is a KL divergence and therefore always greater than 0. this means that once you resolve all the minus signs, the input to the sigmoid is _always_ greater than or equal to 0, which corresponds to a range of P = [0.5, 1], whereas the paper says that for bayesian model selection we should have a range of [0, 1].

Have I missed something here? How does the implementation in the paper "accept ~20% of samples" if P can never be 0? should the sigmoid be some other activation function?

**Comment by tverbele** [comment]
The softmax does not reduce to the sigmoid in this case, as variable u has a dimension of 2: u0 = the old parameters and u1 = the old parameters + the delta in Dirichlet counts should you 'learn' from this sample (see Eq 5). There is indeed no utility in play, but Eq 6 is calculating the mutual information for those both cases and comparing with the softmax (with a very high temperature in the MNIST example). So basically it's saying: only add the Dirichlet counts if the resulting parameters yield higher mutual information between s and o than what it was.

**Reply by Arun-Niranjan** [reply]
Thanks @tverbele that's a bit clearer - I was using `calc_pA_info_gain` which was returning me a single number, I'll give this another look and see if I need a new method or can adapt the existing one with different inputs

**Reply by tverbele** [reply]
you'll need to call that function twice, once with the 'old' pA (or cache that value of a previous call), and once with the new pA which has the Dirichlet counts added; and then check which one is higher / calculate the softmax

**Comment by Arun-Niranjan** [comment]
ok I think I am nearly there with a working classifier (although the best accuracy I am currently getting is ~70%, I'm sure we can do better).

will start loading up my code into a notebook, first PR is here https://github.com/infer-actively/pymdp/pull/357
